lxd/openfga/client.go

```diff
@@ -0,0 +1,174 @@
+package openfga
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+
+	openfga "github.com/openfga/go-sdk"
+	"github.com/openfga/go-sdk/credentials"
+
+	"github.com/canonical/lxd/shared/logger"
+)
+
+type Client struct {
+	apiClient   *openfga.APIClient
+	authModelID string
+}
+
+// NewClient returns a new Client struct containing an OpenFGA APIClient.
+func NewClient(apiURL string, apiToken string, storeID string) (*Client, error) {
+	u, err := url.Parse(apiURL)
+	if err != nil {
+		return nil, fmt.Errorf("Failed parsing URL: %w", err)
+	}
+
+	conf := openfga.Configuration{
+		ApiScheme: u.Scheme,
+		ApiHost:   u.Host,
+		StoreId:   storeID,
+	}
+
+	if apiToken != "" {
+		conf.Credentials = &credentials.Credentials{
+			Method: credentials.CredentialsMethodApiToken,
+			Config: &credentials.Config{
+				ApiToken: apiToken,
+			},
+		}
+	}
+
+	config, err := openfga.NewConfiguration(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client := openfga.NewAPIClient(config)
+
+	// Write authorization model
+	var body openfga.WriteAuthorizationModelRequest
+
+	err = json.Unmarshal([]byte(authModel), &body)
+	if err != nil {
+		return nil, err
+	}
+
+	data, _, err := client.OpenFgaApi.WriteAuthorizationModel(context.Background()).Body(body).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	authModelID := data.GetAuthorizationModelId()
+
+	return &Client{apiClient: client, authModelID: authModelID}, nil
+}
+
+// HasPermission returns true if the user has the relation to the object.
+func (c *Client) HasPermission(user string, relation Relation, object string) bool {
+	fullUser := UserObject(user)
+
+	logger.Debug("Checking ReBAC permission", logger.Ctx{"user": fullUser, "relation": relation, "object": object})
```

Comment:
Can you create a new logger here using `l := logger.AddContext( logger.Ctx{"user": fullUser, "relation": relation, "object": object})`, then you can use `l.Debug()` and include all the context each time, including on line 86.